apollo-federation/src/sources/connect/json_selection/parser.rs

```diff
@@ -346,14 +353,69 @@ impl SubSelection {
         ))(input)
         .map(|(input, (_, _, selections, star, _, _, _))| (input, Self { selections, star }))
     }
+
+    pub fn selections_iter(&self) -> impl Iterator<Item = &NamedSelection> {
+        self.selections.iter()
+    }
+
+    pub fn has_star(&self) -> bool {
+        self.star.is_some()
+    }
+
+    pub fn set_star(&mut self, star: Option<StarSelection>) {
+        self.star = star;
+    }
+
+    pub fn append_selection(&mut self, selection: NamedSelection) {
+        self.selections.push(selection);
+    }
+
+    pub fn last_selection_mut(&mut self) -> Option<&mut NamedSelection> {
+        self.selections.last_mut()
+    }
+
+    // Since we enforce that new selections may only be appended to
+    // self.selections, we can provide an index-based search method that returns
+    // an unforgeable NamedSelectionIndex, which can later be used to access the
+    // selection using either get_at_index or get_at_index_mut.
+    pub fn index_of_named_selection(&self, name: &str) -> Option<NamedSelectionIndex> {
+        self.selections
+            .iter()
+            .position(|selection| selection.name() == name)
+            .map(|pos| NamedSelectionIndex { pos })
+    }
+
+    pub fn get_at_index(&self, index: &NamedSelectionIndex) -> &NamedSelection {
+        self.selections
+            .get(index.pos)
+            .expect("NamedSelectionIndex out of bounds")
+    }
+
+    pub fn get_at_index_mut(&mut self, index: &NamedSelectionIndex) -> &mut NamedSelection {
+        self.selections
+            .get_mut(index.pos)
+            .expect("NamedSelectionIndex out of bounds")
+    }
+}
+
+pub struct NamedSelectionIndex {
+    // Intentionally private so NamedSelectionIndex cannot be forged.
+    pos: usize,
```

Comment:
I think it's pretty cool that `get_at_index` and `get_at_index_mut` can return a `NamedSelection` reference without wrapping it in an `Option`, because of the guarantees provided by `NamedSelectionIndex` and the append-only invariant now enforced by `SubSelection`.